## Single-cell RNAseq and CHiPseq

This project enhance skills in analyzing single cell RNA-seq and CHiP-seq, making use of both Python (Single cell) and Bash (CHiPseq) programming language, along with IGV, iRegulon for assessing regulators, GOrila and GREAT database for GO term enrichment, i-cis-Target for ranking and enriching motifs. <i>Note that this project was running under KU Leuven server on super computer, thus, running this notebook without the correct configuration will result in errors</i>

### Single-cell RNAseq:

In this project, the single-cell RNA-seq analysis is carried out to learn more about the immune system in the skin and blood. The dataset is retrieved from paper: ***Single-cell analysis reveals innate lymphoid cell lineage infidelity in atopic dermatitis*** ([Natalia Alkon et al 2021](https://doi.org/10.1016/j.jaci.2021.07.025)). This study investigated role of innate lymphoid cells (ILCs) by using single-cell RNA-sequencing on skin cells and blood cells from healthy individuals and patients with atopic dermatitis (AD).

The analysis was written in Python language using Jupyter Notebooks. This project was used in the "Regulatory Genomics" course.

### CHiPseq:

In this project, the ChIP-seq analysis is carried out to learn more about ETS1 transcription factor in natural killer cells (NK cell) in human, based on paper ***The transcription factor ETS1 is an important regulator of human NK cell development and terminal differentiation*** ([Sylvie Taveirne et al](https://doi.org/10.1182/blood.2020005204)).

The analysis was written in Bash using Jupyter Notebooks. This project was used in the "Regulatory Genomics" course.